kaitai/kaitaistream.cpp

```diff
@@ -399,11 +399,15 @@ std::string kaitai::kstream::ensure_fixed_contents(std::string expected) {
 }
 
 std::string kaitai::kstream::bytes_strip_right(std::string src, char pad_byte) {
-    std::size_t new_len = src.length();
+    std::size_t max_len = src.length();
+    std::size_t new_len = max_len;
 
     while (new_len > 0 && src[new_len - 1] == pad_byte)
         new_len--;
 
+    if (new_len == max_len)
+        return src;
```

Comment:
I think I understand the problem. In about every other language, both static like C# and dynamically typed like Python Ruby, strings are passed around in O(1). Do I understand correctly that in C++ those strings are passed by full copy in O(N)? If so, then I dont know about the optimisations...

Reply:
C++ offers all three options:
 - pass by `const` reference, where the callee receives a reference (what you're calling O(1)) to the caller's object (here a string), but the callee _may not_ modify it;
 - pass by mutable reference, where the callee receives a reference to the caller's object, and _is_ able to modify it;
 - pass by value, where the callee receives a "copy" of the caller's object -- "copy" is in quotes, because the value is produced using the (possibly default-generated) "copy constructor", which, really, could do anything, including creating a reference to the original object.

However, you have to specify the one that you want...and, the most straightforward syntax produces arguably the safest and simplest effect:  pass by value.

Reply:
Its funny that with all that convoluted answers that you gave me, it makes me think that you actually dont know... which kind of doesnt come as a suprise, I think whoever said "nobodody really understands quantum mechanics" DEFINANETLY meant C++!!

Reply:
I apologize for the confusion.  Which passing mechanism is used is made clear by the function's signature in the declaration/definition.

Pass by "const reference":
```
std::string kaitai::kstream::bytes_strip_right(const std::string& src, char pad_byte);
```
The callee and caller share the storage for the object (here, `src`), but the callee is prevented (at compile-time) from modifying it.  However, it looks, inside the function, like the object was declared in the local scope.  (I.e., `src` is _not_ a pointer, and it is accessed with "dot" notation, as `src.length()`.)

Pass by "reference":
```
std::string kaitai::kstream::bytes_strip_right(std::string& src, char pad_byte);
```
This is the old-fashioned "pass by pointer" hidden under a bunch of syntactic sugar -- the callee and the caller share the storage for the object, the callee accesses it as though it were declared locally, and the callee has the ability to modify it.  This works well for "output parameters" and "in-out parameters", but it is somewhat dangerous, as it's not always apparent to the reader that the caller's value can be modified.

Pass by "value":
```
std::string kaitai::kstream::bytes_strip_right(std::string src, char pad_byte);
```
This one does what it says:  only the "value" of the object gets passed, and the callee receives a copy-constructed version of the object.  By default, that is a full and separate copy of it, which means that the caller is protected from anything that the callee might do, and the callee has full freedom to (ab)use the object as it wishes (once again, the object is accessed as if it were a local declaration); however, there is a performance penalty associated with this if the object is large or complicated to copy.

So, the code as it stands is passing by value, which is arguably the safest, most conservative, least confusing, and probably most expensive approach.

However, that said, it is possible that the compiler can apply some optimizations transparently.  The most obvious is for it to allocate the return value in the caller and use that storage to receive the callee's copy, which removes the need for one of the two copy operations.  (There are other options, depending on what the caller does with the result and on which version of the C++ compiler is being used.)